SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java

```diff
@@ -0,0 +1,175 @@
+import org.w3c.dom.Element;
+
+/**
+ * This class is responsible for authenticating a user given in a SAML Response
+ */
+public final class UserAuthenticator {
+  private static final String REQUEST_SAML_PARAMETER_NAME = "SAMLResponse";
+  private static final X509Certificate SIGN_VERIFICATION_CERT = loadCertificate("myCertificate");
+
+  private final Request request;
+  private Assertion assertion;
+  private Response response;
+
+  public UserAuthenticator(Request request) {
+    this.request = request;
+  }
+
+  /**
+   * Authenticate the user set in the SAML-response POST request parameter.
+   * 
+   * @return the {@link SamlUserAuthenticationResult}
+   * @throws SamlValveException
+   */
+  public UserAuthenticator authenticate() throws AuthenticationException {
+    this.response = unmarshallSamlResponse(decodeSamlResponse(request));
+    validateSamlResponse();
+    return this;
+  }
+
+  public Assertion getAssertion() {
+    return this.assertion;
+  }
+
+  public String getRedirectUri() {
+    return this.response.getDestination();
+  }
+
+  /**
+   * Decodes a SAML response returned by an identity provider.
+   *
+   * @param decodedResponse the encoded response returned by the identity provider.
+   * @return An {@link SamlResponse} object containing information decoded from the SAML response.
+   * @throws SamlException if the signature is invalid, or if any other error occurs.
+   */
+  private static Response unmarshallSamlResponse(String decodedResponse) throws AuthenticationException {
+    DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    documentBuilderFactory.setNamespaceAware(true);
+    try (InputStream is = new ByteArrayInputStream(decodedResponse.getBytes(StandardCharsets.UTF_8))) {
+      BasicParserPool parserMgr = new BasicParserPool();
+      parserMgr.setNamespaceAware(true);
+      Document samlDoc = parserMgr.parse(is);
+      Element samlRoot = samlDoc.getDocumentElement();
+
+      UnmarshallerFactory unmarshallerFactory = Configuration.getUnmarshallerFactory();
+      Unmarshaller unmarshaller = unmarshallerFactory.getUnmarshaller(samlRoot);
+      return (Response) unmarshaller.unmarshall(samlRoot);
+    } catch (XMLParserException | UnmarshallingException | IOException e) {
+      throw new AuthenticationException("Error unmarshalling SAML Response", e);
+    }
+  }
+
+  private void validateSamlResponse() throws AuthenticationException {
+    this.assertion = validateResponse(this.response);
+    validateAssertion(assertion);
+    validateSignature(assertion);
+  }
+
+  private static Assertion validateResponse(Response response) throws AuthenticationException {
+
+    String statusCode = response.getStatus().getStatusCode().getValue();
+    if (!statusCode.equals("urn:oasis:names:tc:SAML:2.0:status:Success")) {
+      throw new AuthenticationException("Invalid status code: " + statusCode);
+    }
+
+    if (response.getAssertions().size() != 1) {

```

Comment:
hm. how does the assertion concept work? who is defining assertions? and what are these assertions? do you have somewhere a doc?

Reply:
The assertion is part of the Identity Provider's Response, and this Response gets created after a user login.
![1127182513244158 Xyqs8g4zWlwULTRe33zB_height640](https://user-images.githubusercontent.com/36727349/78547356-45d25700-77ff-11ea-9ca6-c35d0af5e57e.png)